---
title: "Marrow FACS Notebook"
output:
  html_document: default
  html_notebook: default
---

Enter the directory of the maca folder on your drive and the name of the tissue you want to analyze.

```{r}
tissue_of_interest = "Marrow"
```

Load the requisite packages and some additional helper functions.

```{r}
library(here)
library(useful)
library(Seurat)
library(dplyr)
library(Matrix)
library(ontologyIndex)
cell_ontology = get_ontology('https://raw.githubusercontent.com/obophenotype/cell-ontology/master/cl-basic.obo', extract_tags='everything')

validate_cell_ontology = function(cell_ontology_class){
  in_cell_ontology = sapply(cell_ontology_class, function(x) is.element(x, cell_ontology$name) || is.na(x))
  if (!all(in_cell_ontology)) {
    message = paste0('"', cell_ontology_class[!in_cell_ontology], '" is not in the cell ontology
')
    stop(message)
  }
}
convert_to_cell_ontology_id = function(cell_ontology_class){
  return(sapply(cell_ontology_class, function(x) as.vector(cell_ontology$id[cell_ontology$name == x])[1]))
}

save_dir = here('00_data_ingest', 'tissue_robj')
```



```{r}
# read the metadata to get the plates we want
plate_metadata_filename = here('00_data_ingest', '00_facs_raw_data', 'metadata_FACS.csv')

plate_metadata <- read.csv(plate_metadata_filename, sep=",", header = TRUE)
colnames(plate_metadata)[1] <- "plate.barcode"
plate_metadata
```

Subset the metadata on the tissue.

```{r}
tissue_plates = filter(plate_metadata, tissue == tissue_of_interest)[,c('plate.barcode','tissue','subtissue','mouse.sex')]
tissue_plates
```

Load the read count data.
```{r}
#Load the gene names and set the metadata columns by opening the first file
filename = here('00_data_ingest', '00_facs_raw_data', 'FACS', paste0(tissue_of_interest, '-counts.csv'))

raw.data = read.csv(filename, sep=",", row.names=1)
# raw.data = data.frame(row.names = rownames(raw.data))
corner(raw.data)
```
Make a vector of plate barcodes for each cell

```{r}
plate.barcodes = lapply(colnames(raw.data), function(x) strsplit(strsplit(x, "_")[[1]][1], '.', fixed=TRUE)[[1]][2])
head(plate.barcodes)
```

Use only the metadata rows corresponding to Bladder plates. Make a plate barcode dataframe to "expand" the per-plate metadata to be per-cell.
```{r}
barcode.df = t.data.frame(as.data.frame(plate.barcodes))

rownames(barcode.df) = colnames(raw.data)
colnames(barcode.df) = c('plate.barcode')
head(barcode.df)

rnames = row.names(barcode.df)
meta.data <- merge(barcode.df, plate_metadata, by='plate.barcode', sort = T)
row.names(meta.data) <- rnames
corner(meta.data)

# Sort cells by plate barcode because that's how the data was originally
meta.data = meta.data[order(meta.data$plate.barcode), ]
corner(meta.data)
raw.data = raw.data[, rownames(meta.data)]
corner(raw.data)
```
Process the raw data and load it into the Seurat object.

```{r}
# Find ERCC's, compute the percent ERCC, and drop them from the raw data.
erccs <- grep(pattern = "^ERCC-", x = rownames(x = raw.data), value = TRUE)
percent.ercc <- Matrix::colSums(raw.data[erccs, ])/Matrix::colSums(raw.data)
ercc.index <- grep(pattern = "^ERCC-", x = rownames(x = raw.data), value = FALSE)
raw.data <- raw.data[-ercc.index,]

# Create the Seurat object with all the data
tiss <- CreateSeuratObject(raw.data = raw.data, project = tissue_of_interest, 
                    min.cells = 5, min.genes = 5)

tiss <- AddMetaData(object = tiss, meta.data)
tiss <- AddMetaData(object = tiss, percent.ercc, col.name = "percent.ercc")
# Change default name for sums of counts from nUMI to nReads
colnames(tiss@meta.data)[colnames(tiss@meta.data) == 'nUMI'] <- 'nReads'

# Create metadata columns for cell_ontology_classs and subcell_ontology_classs
tiss@meta.data[,'free_annotation'] <- NA
tiss@meta.data[,'cell_ontology_class'] <- NA
tiss@meta.data[,'subcell_ontology_class'] <- NA
```


Calculate percent ribosomal genes.

```{r}
ribo.genes <- grep(pattern = "^Rp[sl][[:digit:]]", x = rownames(x = tiss@data), value = TRUE)
percent.ribo <- Matrix::colSums(tiss@raw.data[ribo.genes, ])/Matrix::colSums(tiss@raw.data)
tiss <- AddMetaData(object = tiss, metadata = percent.ribo, col.name = "percent.ribo")
```

A sanity check: genes per cell vs reads per cell.

```{r}
GenePlot(object = tiss, gene1 = "nReads", gene2 = "nGene", use.raw=T)
```

Filter out cells with few reads and few genes.

```{r}
tiss <- FilterCells(object = tiss, subset.names = c("nGene", "nReads"), 
    low.thresholds = c(500, 50000), high.thresholds = c(25000, 2000000))
```


Normalize the data, then regress out correlation with total reads
```{r}
tiss <- NormalizeData(object = tiss, scale.factor = 1e6)
tiss <- ScaleData(object = tiss)
tiss <- FindVariableGenes(object = tiss, do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5)
```


Run Principal Component Analysis.
```{r}
tiss <- RunPCA(object = tiss, do.print = FALSE)
tiss <- ProjectPCA(object = tiss, do.print = FALSE)
```

```{r, echo=FALSE, fig.height=4, fig.width=8}
PCHeatmap(object = tiss, pc.use = 1:3, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
```

Later on (in FindClusters and TSNE) you will pick a number of principal components to use. This has the effect of keeping the major directions of variation in the data and, ideally, supressing noise. There is no correct answer to the number to use, but a decent rule of thumb is to go until the plot plateaus.

```{r}
PCElbowPlot(object = tiss)
```

Choose the number of principal components to use.
```{r}
# Set number of principal components. 
n.pcs = 18
```


The clustering is performed based on a nearest neighbors graph. Cells that have similar expression will be joined together. The Louvain algorithm looks for groups of cells with high modularity--more connections within the group than between groups. The resolution parameter determines the scale...higher resolution will give more clusters, lower resolution will give fewer.

For the top-level clustering, aim to under-cluster instead of over-cluster. It will be easy to subset groups and further analyze them below.

```{r}
# Set resolution 
res.used <- 0.5

tiss <- FindClusters(object = tiss, reduction.type = "pca", dims.use = 1:n.pcs, 
    resolution = res.used, print.output = 0, save.SNN = TRUE)
```

To visualize 
```{r}
# If cells are too spread out, you can raise the perplexity. If you have few cells, try a lower perplexity (but never less than 10).
tiss <- RunTSNE(object = tiss, dims.use = 1:n.pcs, seed.use = 10, perplexity=84)
```

```{r}
# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = tiss, do.label = T)
```

Check expression of genes of interset.

```{r, echo=FALSE, fig.height=156, fig.width=15}
genes_to_check = c('Itgam', 'Il7r', 'Kit', 'Atxn1', 'Fcgr3', 'Flt3', 'Cd34', 'Slamf1', 'Gpr56', 'Stmn1', 'Mki67', 'Tmem176b',  'Itgal', 'Itgax', 'Emr1', 'Cd68', 'Cd69', 'Fcgr4', 'Mpeg1', 'Itgb2', 'Ahnak', 'Pld4', 'Cd3e', 'Cd4', 'Cd8a', 'Ly6d', 'Cd27', 'Cr2', 'Fcer2a', 'Cd2', 'Cd7', 'Mme', 'Thy1', 'Klrb1a', 'S100a11', 'Ltf', 'Ngp', 'Fcer1g', 'Pglyrp1', 'Lcn2', 'Camp', 'Hp', 'Ly6g6c', 'Ly6g6e', 'Ptprc', 'Cd19', 'Ms4a1', 'Cox6a2', 'Irf8', 'Cd74', 'Chchd10', 'Cnp', 'Cd79a', 'Cd79b', 'Vpreb1', 'Vpreb3', 'Cd38', 'Cd22', 'Cd24a', 'Cd40', 'Cd48', 'Cd53', 'Cd81', 'Cd84', 'Cxcr4', 'Cxcr5', 'Ccr6', 'Cd200', 'Ly9', 'Abcb1b', 'Tlr9', 'Slamf6', 'H2-Ea-ps', 'Pax5', 'Foxo1', 'Klf4', 'Klf9', 'Mitf', 'Pou2af1', 'Zbtb16', 'Spib', 'Spn', 'Dntt', 'Rag1', 'Rag2', 'Il2ra', 'Pdzk1ip1', 'Ly6a', 'Itga2b', 'Fos', 'Junb', 'Myl10', 'Jun', 'Mn1', 'S100a9', 'Ccl9', 'Tcf3', 'Ebf1', 'Ikzf1', 'Sfpi1', 'Cd1d1', 'Egr2', 'Cd14', 'Mpl', 'Il3ra', 'Bpgm', 'Beta-s', 'Hbb-b2', 'Cpa3', 'Fcer1a', 'Mcpt8', 'Ccl3', 'Gata1', 'Gata2', 'Cebpa')
#genes_to_check = c('Alb', 'Cyp2f2', 'Cyp2e1', 'Hamp')

FeaturePlot(tiss, genes_to_check, pt.size = 1, nCol = 3)
```

Dotplots let you see the intensity of exppression and the fraction of cells expressing for each of your genes of interest.

```{r, echo=FALSE, fig.height=4, fig.width=84}
# To change the y-axis to show raw counts, add use.raw = T.
DotPlot(tiss, genes_to_check, plot.legend = T)
```

How big are the clusters?
```{r}
table(tiss@ident)
```

Which markers identify a specific cluster?

```{r}
clust.markers <- FindMarkers(object = tiss, ident.1 = 0, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
print(x = head(x= clust.markers, n = 10))
```

You can also compute all markers for all clusters at once. This may take some time.
```{r}
tiss.markers <- FindAllMarkers(object = tiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```
```{r}
head(tiss.markers)
```

Display the top markers you computed above.
```{r}
tiss.markers %>% group_by(cluster) %>% top_n(5, avg_logFC)
```

## Assigning cell type identity to clusters

At a coarse level, we can use canonical markers to match the unbiased clustering to known cell types:

0: alpha
1: beta
2: beta
3: exocrine
4: duct
5: delta
6: gamma
7: endothelial
8: immune
9: stellate

```{r}
# stash current cluster IDs
tiss <- StashIdent(object = tiss, save.name = "cluster.ids")

# enumerate current cluster IDs and the labels for them
cluster.ids <- c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13)
free_annotation <- c()
cell_ontology_class <-c(
  "hematopoietic precursor cell", 
  "naive B cell", 
  "pre-B cell", 
  "granulocyte", 
  "granulocyte", 
  "immature B cell", 
  "late pro-B cell", 
  "T_NK_NKT cell", 
  "monocyte", 
  "granulocytopoietic cell", 
  "hematopoietic precursor cell", 
  "macrophage", 
  "granulocyte monocyte progenitor cell", 
  "basophil")
cell_ontology_id <- c(
  "CL:0008001", 
  "CL:0000788", 
  "CL:0000817", 
  "CL:0000094", 
  "CL:0000094", 
  "CL:0000816", 
  "CL:0002048", 
  NA, 
  "CL:0000576", 
  "CL:0002191", 
  "CL:0008001", 
  "CL:0000235", 
  "CL:0000557", 
  "CL:0000767")
  

tiss@meta.data[,'free_annotation'] <- NA
tiss@meta.data[,'cell_ontology_class'] <- plyr::mapvalues(x = tiss@ident, from = cluster.ids, to = cell_ontology_class)
tiss@meta.data[,'cell_ontology_id'] <- plyr::mapvalues(x = tiss@ident, from = cluster.ids, to = cell_ontology_id)

tiss@meta.data[tiss@cell.names,'cell_ontology_class'] <- as.character(tiss@meta.data$cell_ontology_class)
tiss@meta.data[tiss@cell.names,'cell_ontology_id'] <- as.character(tiss@meta.data$cell_ontology_id)


TSNEPlot(object = tiss, do.label = TRUE, pt.size = 0.5, group.by='cell_ontology_class')
```


## Checking for batch effects


Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = tiss, do.return = TRUE, group.by = "plate.barcode")
```

Print a table showing the count of cells in each identity category from each plate.

```{r}
table(as.character(tiss@ident), as.character(tiss@meta.data$plate.barcode))
```

## Redo Subcluster as per the code from pancreas plate Rmd, which is without regressing out ribosomal genes
## 

####################################
#####################################
######################################


## Subset == T_NK_NKT cell (Cluster 7)

```{r}
subtiss = SubsetData(tiss, ident.use = c(7))
```

```{r}
subtiss <- subtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subtiss, pc.use = 1:3, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subtiss)
```


```{r}
sub.n.pcs = 4
sub.res.use = 1
subtiss <- subtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:sub.n.pcs, 
    resolution = sub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:sub.n.pcs, seed.use = 10, perplexity=44)

TSNEPlot(object = subtiss, do.label = T, pt.size = 1.2, label.size = 4)

```
Check expression of genes of interset.

```{r}
subtiss.markers <- FindAllMarkers(object = subtiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subtiss.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=102, fig.width=15}
genes_to_check = c('Cd6','Il7r','Ctla4','Cd8b1', 'Cd69', 'Cxcr6', 'Cd4', 'Cd8a', 'Cd8b1', 'Ccr7', 'Tcf7', 'Lef1', 'Mmp9', 'Tnfrsf4', 'Foxp3', 'Lcn2', 'Klrc1', 'Cd160', 'Ctla2a', 'Tyrobp', 'Ncr1', 'Cd3e', 'Klrb1a', 'Klrb1c', 'Prf1', 'Serpinb9', 'Lyz2', 'Ngp', 'Hp', 'Ly6c2', 'Cd79a', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Top2a', 'Rrm2', 'Nkg7', 'Cd1d1', 'Egr2', 'Cd19', 'Ms4a1', 'Chchd10', 'Cnp', 'Rag1', 'Rag2', 'Dntt', 'Pax5', 'Il2ra', 'Cxcr5', 'Ccr6', 'Cr2', 'Cd22', 'Vpreb3', 'Mki67', 'Stmn1', 'Il2rb', 'Pld4', 'Cd68', 'Mpeg1', 'Emr1', 'Adamts14', 'Itgax', 'Styk1', 'Ugt1a7c', 'Khdc1a', 'Car5b', 'Gzma', 'Cma1', 'A430084P05Rik', 'Ccl4', 'Sh2d1b1', 'Ncam1', 'Klra1', 'Cd34', 'Il2rb', 'Itga4', 'Itgb7')

FeaturePlot(subtiss, genes_to_check, pt.size = 1, nCol = 3)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subtiss, c('mouse.id','ident')) %>% droplevels())
```

```{r}
sub.cluster.ids <- c(0, 1, 2, 3, 4, 5)
sub.free_annotation <- c("regulatory_immature T cell", "mature natural killer cell", "immature NK T cell", "Cd3e+_Klrb1+ B cell", "immature natural killer cell", "pre-natural killer cell")
#sub.free_annotation <- c()
sub.cell_ontology_id = c(NA,
  "CL:0000824",
  "CL:0000914",
  NA,
  "CL:0000823",
  "CL:0000937")
sub.cell_ontology_class <-c(NA, "mature natural killer cell", "immature NK T cell", NA, "immature natural killer cell", "pre-natural killer cell")

validate_cell_ontology(sub.cell_ontology_class)
sub.cell_ontology_id = convert_to_cell_ontology_id(sub.cell_ontology_class)

subtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.free_annotation))
validate_cell_ontology(sub.cell_ontology_class)
sub.cell_ontology_id = convert_to_cell_ontology_id(sub.cell_ontology_class)

subtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.free_annotation))
subtiss@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.cell_ontology_class))
subtiss@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.cell_ontology_id))
```

## Add subcluster cell_ontology_classs to main cell_ontology_class

```{r}
sub.cells = rownames(subtiss@meta.data)

tiss@meta.data[sub.cells, 'free_annotation'] = subtiss@meta.data[,'free_annotation']
tiss@meta.data[sub.cells, 'cell_ontology_class'] = subtiss@meta.data[,'cell_ontology_class']
tiss@meta.data[sub.cells, 'cell_ontology_id'] = subtiss@meta.data[,'cell_ontology_id']
```


## Checking for batch effects

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subtiss, do.return = TRUE, group.by = "plate.barcode")
```

# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subtiss, do.label = TRUE, do.return = TRUE, group.by = "free_annotation")
```

# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

######################################################
#######################################################

## Multiple cell-types in the cluster 0 of subtiss (original cluster 7)
## subcluster the cluster 0 (regulatory_immature T cell) of subtiss (original cluster 7)

## Subset == regulatory_immature T cell (Cluster 0)

```{r}
subsubtiss = SubsetData(subtiss, ident.use = c(0))
```

```{r}
subsubtiss <- subsubtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subsubtiss, pc.use = 1:3, cells.use = 170, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subsubtiss)
```


```{r}
subsub.n.pcs = 3
subsub.res.use = 1
subsubtiss <- subsubtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:subsub.n.pcs, 
    resolution = subsub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:subsub.n.pcs, seed.use = 10, perplexity=12)

TSNEPlot(object = subsubtiss, do.label = T, pt.size = 1.2, label.size = 4)

```
Check expression of genes of interset.

```{r}
subsubtiss.markers <- FindAllMarkers(object = subsubtiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subsubtiss.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=84, fig.width=15}
genes_to_check = c('Cd6','Il7r','Ctla4','Cd8b1', 'Cd69', 'Cxcr6', 'Cd4', 'Cd8a', 'Cd8b1', 'Ccr7', 'Tcf7', 'Lef1', 'Mmp9', 'Tnfrsf4', 'Foxp3', 'Lcn2', 'Klrc1', 'Cd160', 'Ctla2a', 'Tyrobp', 'Cd3e', 'Klrb1a', 'Klrb1c', 'Prf1', 'Serpinb9', 'Lyz2', 'Ngp', 'Hp', 'Ly6c2', 'Cd79a', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Top2a', 'Rrm2', 'Nkg7', 'Cd1d1', 'Egr2', 'Cd19', 'Ms4a1', 'Chchd10', 'Cnp', 'Rag2', 'Dntt', 'Il2ra', 'Cxcr5', 'Ccr6', 'Cr2', 'Cd22', 'Vpreb3', 'Mki67', 'Stmn1', 'Il2rb', 'Pld4', 'Cd68', 'Mpeg1', 'Ugt1a7c', 'Car5b', 'Gzma', 'A430084P05Rik', 'Ccl4', 'Ncam1', 'Klra1', 'Cd34', 'Itga4', 'Itgb7', 'Tgfb1', 'Il10')

FeaturePlot(subsubtiss, genes_to_check, pt.size = 1, nCol = 3)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subsubtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subsubtiss, c('mouse.id','ident')) %>% droplevels())
```

```{r}
subsub.cluster.ids <- c(0, 1, 2)
subsub.free_annotation <- c(
  "regulatory T cell", 
  "immature T cell", 
  "immature T cell")
#subsub.free_annotation <- c()
subsub.cell_ontology_id = c(
  "CL:0000815",
  "CL:0002420",
  "CL:0002420")
subsub.cell_ontology_class <-c("regulatory T cell", "immature T cell", "immature T cell")

validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
subsubtiss@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_class))
subsubtiss@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_id))
```

## Add subsubcluster cell_ontology_classs to main cell_ontology_class

```{r}
subsub.cells = rownames(subsubtiss@meta.data)

tiss@meta.data[subsub.cells, 'free_annotation'] = subsubtiss@meta.data[,'free_annotation']
tiss@meta.data[subsub.cells, 'cell_ontology_class'] = subsubtiss@meta.data[,'cell_ontology_class']
tiss@meta.data[subsub.cells, 'cell_ontology_id'] = subsubtiss@meta.data[,'cell_ontology_id']
```


## Checking for batch effects

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subsubtiss, do.return = TRUE, group.by = "plate.barcode")
```

# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subsubtiss, do.label = TRUE, do.return = TRUE, group.by = "cell_ontology_class")
```

# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

####################################
#####################################
######################################


## Subset == late pro-B cell (Cluster 6)

## late pro-B cells are reportedly committed to B cell lineage as they begin expressing Pax5. They also have Rag1, Rag2, and Dntt expression. At single cell resolution there seem to be two subsets of late pro-B cells: Dntt+ and Dntt-.

## It appears there are multiple novel cell-types within the Pax5+ late pro-B cell

```{r}
subtiss = SubsetData(tiss, ident.use = c(6))
```

```{r}
subtiss <- subtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subtiss, pc.use = 1:3, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subtiss)
```


```{r}
sub.n.pcs = 4
sub.res.use = 1
subtiss <- subtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:sub.n.pcs, 
    resolution = sub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:sub.n.pcs, seed.use = 10, perplexity=42)

TSNEPlot(object = subtiss, do.label = T, pt.size = 1.2, label.size = 4)

```
Check expression of genes of interset.

```{r}
subtiss.markers <- FindAllMarkers(object = subtiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subtiss.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=102, fig.width=15}
genes_to_check = c('Il7r', 'Cd69', 'Lyz2', 'Tbxa2r', 'Vpreb1', 'Vpreb2', 'Igll1', 'Vpreb3', 'Cd79a', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Stmn1', 'Top2a', 'Mki67', 'Rrm2', 'Nkg7', 'Cd1d1', 'Egr2', 'Cd19', 'Ms4a1', 'Chchd10', 'Cnp', 'Rag1', 'Rag2', 'Dntt', 'Pax5', 'Il2ra', 'Cxcr5', 'Ccr6', 'Cr2', 'Cd22', 'Vpreb3', 'Mki67', 'Stmn1', 'Il2rb', 'Cdc20', 'Ube2c', 'Ccnb1', 'Cenpf', 'Kpna2', 'Tcf3', 'Ebf1', 'Ikzf1', 'Sfpi1', 'Slco4a1', 'Zfp810', 'Ska1', 'Kif18a', 'Depdc1b', 'Btg2', 'S100a8', 'Serinc5', 'Srm', 'Smyd2', 'Comtd1', 'Erg', 'Arpp21', 'Mmrn1', 'Kdm5b', 'Grb7', 'Acy3', 'Ung', 'Cybb', 'Klhl15', 'E2f1', 'Troap', 'Eri2', 'Ltb')

FeaturePlot(subtiss, genes_to_check, pt.size = 1, nCol = 3)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subtiss, c('mouse.id','ident')) %>% droplevels())
```

```{r}
sub.cluster.ids <- c(0, 1, 2, 3, 4)
sub.free_annotation <- c("Dntt- late pro-B cell", "Dntt+ late pro-B cell", "Dntt- late pro-B cell", "Dntt- late pro-B cell", "Dntt+ late pro-B cell")
#sub.free_annotation <- c()
sub.cell_ontology_id = c(NA,
  NA,
  NA,
  NA,
  NA)
sub.cell_ontology_class <-c(NA, NA, NA, NA, NA)

validate_cell_ontology(sub.cell_ontology_class)
sub.cell_ontology_id = convert_to_cell_ontology_id(sub.cell_ontology_class)

subtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.free_annotation))
validate_cell_ontology(sub.cell_ontology_class)
sub.cell_ontology_id = convert_to_cell_ontology_id(sub.cell_ontology_class)

subtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.free_annotation))
subtiss@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.cell_ontology_class))
subtiss@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.cell_ontology_id))
```

## Add subcluster cell_ontology_classs to main cell_ontology_class

```{r}
sub.cells = rownames(subtiss@meta.data)

tiss@meta.data[sub.cells, 'free_annotation'] = subtiss@meta.data[,'free_annotation']
tiss@meta.data[sub.cells, 'cell_ontology_class'] = subtiss@meta.data[,'cell_ontology_class']
tiss@meta.data[sub.cells, 'cell_ontology_id'] = subtiss@meta.data[,'cell_ontology_id']
```


## Checking for batch effects

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subtiss, do.return = TRUE, group.by = "plate.barcode")
```

# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subtiss, do.label = TRUE, do.return = TRUE, group.by = "free_annotation")
```

# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

####################################
#####################################
######################################


## Subset == hematopoietic precursor cells (Cluster 0 and 10)
## This subset contains hematopoietic multipotent progenitor (MPP) cells and a very minute fraction of Hematopoietic stem cells (HSC).


```{r}
subtiss = SubsetData(tiss, ident.use = c(0, 10))
```

```{r}
subtiss <- subtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subtiss, pc.use = 1:3, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subtiss)
```


```{r}
sub.n.pcs = 20
sub.res.use = 1
subtiss <- subtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:sub.n.pcs, 
    resolution = sub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:sub.n.pcs, seed.use = 10, perplexity=65)

TSNEPlot(object = subtiss, do.label = T, pt.size = 1.2, label.size = 4)
```
Check expression of genes of interset.

```{r}
subtiss.markers <- FindAllMarkers(object = subtiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subtiss.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=120, fig.width=15}
genes_to_check = c('Il7r', 'Kit', 'Atxn1', 'Cd34', 'Slamf1', 'Flt3', 'Cd63', 'Myl12b', 'Muc13', 'Sdsl', 'Stat3', 'Wtap', 'Mpl', 'Txnip', 'Ltb', 'Pdzk1ip1', 'Esam', 'Cish', 'Tyms', 'Tk1', '2810417H13Rik', 'Rrm2', 'Uhrf1', 'Tuba1b', 'Myl10', 'Junb', 'Fos', 'Ier2', 'Dusp1', 'Ly6a', 'Ube2c', 'Ccnb2', 'Ccnb1', 'Cdc20', 'Ccna2', 'Flt3', 'Dntt', 'Gm11428', 'Ncf1', 'Mn1', 'H2-Ob', 'Ctsg', 'Mpo', 'Ccl9', 'Sell', 'Hk3', 'Cd48', 'Gata1', 'Mfsd2b', 'Tuba8', 'Itga2b', 'Apoe', 'Pf4', 'Elane', 'Ms4a3', 'Ctsg', 'Mpo', 'Prtn3', 'Cd69', 'Ccr7', 'Ctla2a', 'Tyrobp', 'Serpinb9', 'Lyz2', 'Hp', 'Ly6c2', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Top2a', 'Rrm2', 'Nkg7', 'Cd1d1', 'Chchd10', 'Cnp', 'Rag2', 'Dntt', 'Mki67', 'Stmn1', 'Gimap8', 'Jhdm1d', 'Kif11', 'Spag5', 'Xist', 'Tsix', 'Notch2', 'Slc39a1', 'Socs3', 'Vldlr', 'Mcm4', 'Ung')

FeaturePlot(subtiss, genes_to_check, pt.size = 1, nCol = 3)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subtiss, c('mouse.id','ident')) %>% droplevels())
```

```{r}
sub.cluster.ids <- c(0, 1, 2, 3, 4, 5, 6, 7)
sub.free_annotation <- c("hematopoietic precursor cell", "MAA000848", "hematopoietic precursor cell", "MAA001844", "MAA001887", "hematopoietic precursor cell", "hematopoietic precursor cell", "hematopoietic precursor cell")
#sub.free_annotation <- c()
sub.cell_ontology_id = c("CL:0008001",
  NA,
  "CL:0008001",
  NA,
  NA,
  "CL:0008001",
  "CL:0008001", 
  "CL:0008001")
sub.cell_ontology_class <-c("hematopoietic precursor cell",
  NA,
  "hematopoietic precursor cell",
  NA,
  NA,
  "hematopoietic precursor cell",
  "hematopoietic precursor cell", 
  "hematopoietic precursor cell")
validate_cell_ontology(sub.cell_ontology_class)
sub.cell_ontology_id = convert_to_cell_ontology_id(sub.cell_ontology_class)

subtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.free_annotation))
validate_cell_ontology(sub.cell_ontology_class)
sub.cell_ontology_id = convert_to_cell_ontology_id(sub.cell_ontology_class)

subtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.free_annotation))
subtiss@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.cell_ontology_class))
subtiss@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subtiss@ident, from = sub.cluster.ids, to = sub.cell_ontology_id))
```

## Add subcluster cell_ontology_classs to main cell_ontology_class

```{r}
sub.cells = rownames(subtiss@meta.data)

tiss@meta.data[sub.cells, 'free_annotation'] = subtiss@meta.data[,'free_annotation']
tiss@meta.data[sub.cells, 'cell_ontology_class'] = subtiss@meta.data[,'cell_ontology_class']
tiss@meta.data[sub.cells, 'cell_ontology_id'] = subtiss@meta.data[,'cell_ontology_id']
```


## Checking for batch effects

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subtiss, do.return = TRUE, do.label = TRUE, group.by = "plate.barcode")
```

# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subtiss, do.label = TRUE, do.return = TRUE, group.by = "free_annotation")
```

plate.barcode,mouse.id,tissue,subtissue,FACS.selection,mouse.sex
MAA000848,3_10_M,Marrow,KLS,Multiple,M
MAA001844,3_38_F,Marrow,KLS,Multiple,F
MAA001887,3_39_F,Marrow,KLS,Multiple,F
# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

######################################################
#######################################################

## Remove plate specific cells affected by batch effect.
## Subcluster the clusters 0, 2, 5, 6, and 7 (hematopoietic precursor cell) of subtiss (original cluster 0, and 10)

## Subset == hematopoietic precursor cell (Cluster 0, 2, 5, 6, and 7)

```{r}
subsubtiss = SubsetData(subtiss, ident.use = c(0, 2, 5, 6, 7))
```

```{r}
subsubtiss <- subsubtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subsubtiss, pc.use = 1:3, cells.use = 170, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subsubtiss)
```

```{r}
subsub.n.pcs = 19
subsub.res.use = 1
subsubtiss <- subsubtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:subsub.n.pcs, 
    resolution = subsub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:subsub.n.pcs, seed.use = 10, perplexity=60)

TSNEPlot(object = subsubtiss, do.label = T, pt.size = 1.2, label.size = 4)
```


```{r}
subsub.n.pcs = 20
subsub.res.use = 1
subsubtiss <- subsubtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:subsub.n.pcs, 
    resolution = subsub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:subsub.n.pcs, seed.use = 10, perplexity=60)

TSNEPlot(object = subsubtiss, do.label = T, pt.size = 1.2, label.size = 4)

```
Check expression of genes of interset.

```{r}
subsubtiss.markers <- FindAllMarkers(object = subsubtiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subsubtiss.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=138, fig.width=15}
genes_to_check = c('Hoxb5', 'Trib3', 'Neo1', 'Il7r', 'Kit', 'Atxn1', 'Cd34', 'Slamf1', 'Flt3', 'Cd63', 'Myl12b', 'Muc13', 'Sdsl', 'Stat3', 'Wtap', 'Mpl', 'Txnip', 'Ltb', 'Pdzk1ip1', 'Esam', 'Cish', 'Tyms', 'Tk1', '2810417H13Rik', 'Rrm2', 'Uhrf1', 'Tuba1b', 'Myl10', 'Junb', 'Fos', 'Ier2', 'Dusp1', 'Ly6a', 'Ube2c', 'Ccnb2', 'Ccnb1', 'Cdc20', 'Ccna2', 'Dntt', 'Gm11428', 'Ncf1', 'Mn1', 'H2-Ob', 'Ctsg', 'Mpo', 'Ccl9', 'Sell', 'Hk3', 'Cd48', 'Gata1', 'Mfsd2b', 'Tuba8', 'Itga2b', 'Apoe', 'Pf4', 'Elane', 'Ms4a3', 'Ctsg', 'Prtn3', 'Cd69', 'Ccr7', 'Ctla2a', 'Tyrobp', 'Serpinb9', 'Lyz2', 'Hp', 'Ly6c2', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Top2a', 'Nkg7', 'Cd1d1', 'Chchd10', 'Cnp', 'Rag1',  'Rag2', 'Dntt', 'Mki67', 'Stmn1', 'Gimap8', 'Jhdm1d', 'Kif11', 'Spag5', 'Xist', 'Tsix', 'Notch2', 'Slc39a1', 'Socs3', 'Vldlr', 'Mcm4', 'Ung', 'Fam102a', 'Ccl4', 'Pyy', 'Ak4',  'Prc1', 'Cenpf', 'Prr11', 'Lhcgr', 'Tgm2', 'Abcg3', 'Ccne2', 'Pask', 'Nsg1', 'Bok', 'Klf1', 'Treml1')

FeaturePlot(subsubtiss, genes_to_check, pt.size = 1, nCol = 3)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subsubtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subsubtiss, c('mouse.id','ident')) %>% droplevels())
```

```{r}
subsub.cluster.ids <- c(0, 1, 2, 3, 4, 5, 6)
subsub.free_annotation <- c(
  "Slamf1-negative multipotent progenitor cell", 
  "Slamf1-negative multipotent progenitor cell", 
  "common lymphoid progenitor", 
  "HSC_MPPa", 
  "Slamf1-positive multipotent progenitor cell", 
  "granulocyte monocyte progenitor cell", 
  "megakaryocyte-erythroid progenitor cell")
#subsub.free_annotation <- c()
subsub.cell_ontology_id = c(
  "CL:0002035",
  "CL:0002035",
  "CL:0000051", 
  NA, 
  "CL:0002036", 
  "CL:0000557", 
  "CL:0000050")
subsub.cell_ontology_class <-c("regulatory T cell", "Slamf1-negative multipotent progenitor cell", "common lymphoid progenitor", NA, "Slamf1-positive multipotent progenitor cell", "granulocyte monocyte progenitor cell", "megakaryocyte-erythroid progenitor cell")

validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
subsubtiss@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_class))
subsubtiss@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subsubtiss@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_id))
```

## Add subsubcluster cell_ontology_classs to main cell_ontology_class

```{r}
subsub.cells = rownames(subsubtiss@meta.data)

tiss@meta.data[subsub.cells, 'free_annotation'] = subsubtiss@meta.data[,'free_annotation']
tiss@meta.data[subsub.cells, 'cell_ontology_class'] = subsubtiss@meta.data[,'cell_ontology_class']
tiss@meta.data[subsub.cells, 'cell_ontology_id'] = subsubtiss@meta.data[,'cell_ontology_id']
```


## Checking for batch effects
plate.barcode,mouse.id,tissue,subtissue,FACS.selection,mouse.sex
D042044,3_9_M,Marrow,B-cells,Multiple,M
D042467,3_9_M,Marrow,T-cells,Multiple,M
D041912,3_8_M,Marrow,Granulocytes,Multiple,M
D042193,3_8_M,Marrow,T-cells,Multiple,M
D042479,3_8_M,Marrow,B-cells,Multiple,M
MAA000409,3_10_M,Marrow,Granulocytes,Multiple,M
MAA000596,3_9_M,Marrow,KLS,Multiple,M
MAA000600,3_8_M,Marrow,KLS,Multiple,M
MAA000639,3_9_M,Marrow,Granulocytes,Multiple,M
MAA000652,3_10_M,Marrow,B-cells,Multiple,M
MAA000844,3_10_M,Marrow,T-cells,Multiple,M
MAA000848,3_10_M,Marrow,KLS,Multiple,M
D045139,3_39_F,Marrow,Granulocytes,Multiple,F
D045140,3_38_F,Marrow,Granulocytes,Multiple,F
MAA001842,3_38_F,Marrow,T-cells,Multiple,F
MAA001844,3_38_F,Marrow,KLS,Multiple,F
MAA001883,3_39_F,Marrow,B-cells,Multiple,F
MAA001884,3_38_F,Marrow,B-cells,Multiple,F
MAA001887,3_39_F,Marrow,KLS,Multiple,F
MAA001888,3_39_F,Marrow,T-cells,Multiple,F

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subsubtiss, do.return = TRUE, group.by = "plate.barcode")
```

# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subsubtiss, do.label = TRUE, do.return = TRUE, group.by = "free_annotation")
```

# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

```{r}
filename = here('00_data_ingest', '03_tissue_annotation_csv', 
                     paste0(tissue_of_interest, "_facs_annotation.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```


######################################################
#######################################################

## Try to resolve HSCs and MPPa.

## Subset == HSC_MPPa (Cluster 3) of cluster hematopoietic precursor cell (clusters 0, 2, 5, 6, and 7) of original hematopoietic precursor cell (clusters 0, and 10)

```{r}
subsubsubtiss = SubsetData(subsubtiss, ident.use = c(3))
```

```{r}
subsubsubtiss <- subsubsubtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subsubsubtiss, pc.use = 1:3, cells.use = 170, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subsubsubtiss)
```


```{r}
subsubsub.n.pcs = 5
subsub.res.use = 1
subsubsubtiss <- subsubsubtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:subsubsub.n.pcs, 
    resolution = subsub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:subsubsub.n.pcs, seed.use = 10, perplexity=21)

TSNEPlot(object = subsubsubtiss, do.label = T, pt.size = 1.2, label.size = 4)

```
Check expression of genes of interset.

```{r}
subsubsubtiss.markers <- FindAllMarkers(object = subsubsubtiss, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subsubsubtiss.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=102, fig.width=16}
genes_to_check = c('Hoxb5', 'Trib3', 'Neo1', 'Kit', 'Atxn1', 'Cd34', 'Slamf1', 'Flt3', 'Cd63', 'Myl12b', 'Muc13', 'Sdsl', 'Stat3', 'Wtap', 'Mpl', 'Txnip', 'Ltb', 'Pdzk1ip1', 'Esam', 'Cish', 'Tyms', 'Tk1', '2810417H13Rik', 'Rrm2', 'Uhrf1', 'Tuba1b', 'Myl10', 'Junb', 'Fos', 'Ier2', 'Dusp1', 'Ly6a', 'Ube2c', 'Ccnb2', 'Ccnb1', 'Cdc20', 'Ccna2', 'Dntt', 'Gm11428', 'Ncf1', 'Mn1', 'H2-Ob', 'Ctsg', 'Mpo', 'Ccl9', 'Sell', 'Hk3', 'Cd48', 'Gata1', 'Mfsd2b', 'Tuba8', 'Itga2b', 'Apoe', 'Pf4', 'Elane', 'Ctsg', 'Prtn3', 'Cd69', 'Ccr7', 'Ctla2a', 'Tyrobp', 'Serpinb9', 'Lyz2', 'Hp', 'Ly6c2', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Top2a', 'Nkg7', 'Cd1d1', 'Chchd10', 'Cnp', 'Rag1',  'Rag2', 'Dntt', 'Mki67', 'Stmn1', 'Gimap8', 'Jhdm1d', 'Kif11', 'Spag5', 'Xist', 'Tsix', 'Notch2', 'Slc39a1', 'Socs3', 'Vldlr', 'Mcm4', 'Ung', 'Fam102a', 'Ccl4', 'Pyy', 'Ak4',  'Prc1', 'Cenpf', 'Prr11', 'Lhcgr', 'Tgm2', 'Abcg3', 'Ccne2', 'Pask', 'Nsg1', 'Bok', 'Klf1', 'Treml1', 'Gfi1', 'Ccl9', 'Dyrk1b', 'Sdc1', 'Axl', 'Itgb7', 'Sult1a1', 'Plcl1', 'Cdkn1c', 'Slfn2', 'Ankrd13c', 'Tenc1', 'Trmt5', '4921511C10Rik', 'Heatr7a', 'Cdc6', 'Scfd2', 'Mrgpre')

FeaturePlot(subsubsubtiss, genes_to_check, pt.size = 1, nCol = 4)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subsubsubtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subsubsubtiss, c('mouse.id','ident')) %>% droplevels())
```

```{r}
subsub.cluster.ids <- c(0, 1)
subsub.free_annotation <- c(
  "hematopoietic stem cell enriched fraction", 
  "Slamf1-negative multipotent progenitor cell")
#subsub.free_annotation <- c()
subsub.cell_ontology_id = c(
  "CL:0000037",
  "CL:0002035")
subsub.cell_ontology_class <-c("hematopoietic stem cell", "Slamf1-negative multipotent progenitor cell")

validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubsubtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubsubtiss@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubsubtiss@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubsubtiss@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
subsubsubtiss@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subsubsubtiss@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_class))
subsubsubtiss@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subsubsubtiss@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_id))
```

## Add subsubcluster cell_ontology_classs to main cell_ontology_class

```{r}
subsub.cells = rownames(subsubsubtiss@meta.data)

tiss@meta.data[subsub.cells, 'free_annotation'] = subsubsubtiss@meta.data[,'free_annotation']
tiss@meta.data[subsub.cells, 'cell_ontology_class'] = subsubsubtiss@meta.data[,'cell_ontology_class']
tiss@meta.data[subsub.cells, 'cell_ontology_id'] = subsubsubtiss@meta.data[,'cell_ontology_id']
```


## Checking for batch effects

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subsubsubtiss, do.return = TRUE, group.by = "plate.barcode")
```

# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subsubsubtiss, do.label = TRUE, do.return = TRUE, group.by = "free_annotation")
```

# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

```{r}
filename = here('00_data_ingest', '03_tissue_annotation_csv', 
                     paste0(tissue_of_interest, "_facs_annotation.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

######################################################
#######################################################

## Try to see if the excluded cells that were affected by batch effect contain any known cell-types.
## As these cells were originally sorted in the KLS plate they should ideally only contain HSCs and MPPs
## Subcluster the clusters 1, 4, and 3 (Cells affected by batch-effect) of subtiss (original cluster 0, and 10)

## Subset == MAA000848 (KLS), MAA001844 (KLS), and MAA00188 (KLS) (Cluster 1, 3, 4) of original hematopoietic precursor cell (clusters 0, and 10)

```{r}
subsubtiss = SubsetData(subtiss, ident.use = c(1, 4))
```

```{r}
subsubtiss <- subsubtiss %>% ScaleData() %>% 
  FindVariableGenes(do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5) %>%
  RunPCA(do.print = FALSE)
```

```{r}
PCHeatmap(object = subsubtiss, pc.use = 1:3, cells.use = 170, do.balanced = TRUE, label.columns = FALSE, num.genes = 8)
PCElbowPlot(subsubtiss)
```

# Find out the PC that is picking up the batch effect by plate.

```{r}
PCAPlot(object = subsubtiss, do.return = TRUE, dim.1 = 1, dim.2 = 2, group.by = "plate.barcode")
PCAPlot(object = subsubtiss, do.return = TRUE, dim.1 = 1, dim.2 = 3, group.by = "plate.barcode")
PCAPlot(object = subsubtiss, do.return = TRUE, dim.1 = 2, dim.2 = 3, group.by = "plate.barcode")
```

# Regress out PC2
# Run PCA again

```{r}
subsubtiss1 <- ScaleData(object = subsubtiss, vars.to.regress = "PC2")
subsubtiss1 <- RunPCA(object = subsubtiss1,  do.print = FALSE, pcs.print = 5, genes.print = 5)
```


```{r}
PCAPlot(object = subsubtiss1, do.return = TRUE, group.by = "plate.barcode")
PCAPlot(object = subsubtiss1, do.return = TRUE, dim.1 = 1, dim.2 = 3, group.by = "plate.barcode")
PCAPlot(object = subsubtiss1, do.return = TRUE, dim.1 = 2, dim.2 = 3, group.by = "plate.barcode")
```

```{r}
PCElbowPlot(subsubtiss1)
```


```{r}
subsub.n.pcs = 15
subsub.res.use = 1
subsubtiss1 <- subsubtiss1 %>% FindClusters(reduction.type = "pca", dims.use = 1:subsub.n.pcs, 
    resolution = subsub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:subsub.n.pcs, seed.use = 10, perplexity=60)

TSNEPlot(object = subsubtiss1, do.label = T, pt.size = 1.2, label.size = 4)

```


```{r}
subsub.n.pcs = 15
subsub.res.use = 1
subsubtiss <- subsubtiss %>% FindClusters(reduction.type = "pca", dims.use = 1:subsub.n.pcs, 
    resolution = subsub.res.use, print.output = 0, save.SNN = TRUE) %>%
    RunTSNE(dims.use = 1:subsub.n.pcs, seed.use = 10, perplexity=60)

TSNEPlot(object = subsubtiss, do.label = T, pt.size = 1.2, label.size = 4)
```


# Check expression of genes of interset.

```{r}
subsubtiss1.markers <- FindAllMarkers(object = subsubtiss1, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
```

```{r}
subsubtiss1.markers %>% group_by(cluster) %>% top_n(6, avg_logFC)
```



```{r, echo=FALSE, fig.height=138, fig.width=15}
genes_to_check = c('Hoxb5', 'Trib3', 'Neo1', 'Il7r', 'Kit', 'Atxn1', 'Cd34', 'Slamf1', 'Flt3', 'Cd63', 'Myl12b', 'Muc13', 'Sdsl', 'Stat3', 'Wtap', 'Mpl', 'Txnip', 'Ltb', 'Pdzk1ip1', 'Esam', 'Cish', 'Tyms', 'Tk1', '2810417H13Rik', 'Rrm2', 'Uhrf1', 'Tuba1b', 'Myl10', 'Junb', 'Fos', 'Ier2', 'Dusp1', 'Ly6a', 'Ube2c', 'Ccnb2', 'Ccnb1', 'Cdc20', 'Ccna2', 'Dntt', 'Gm11428', 'Ncf1', 'Mn1', 'H2-Ob', 'Ctsg', 'Mpo', 'Ccl9', 'Sell', 'Hk3', 'Cd48', 'Gata1', 'Mfsd2b', 'Tuba8', 'Itga2b', 'Apoe', 'Pf4', 'Elane', 'Ms4a3', 'Ctsg', 'Prtn3', 'Cd69', 'Ccr7', 'Ctla2a', 'Tyrobp', 'Serpinb9', 'Lyz2', 'Hp', 'Ly6c2', 'Cd74', 'H2-Aa', 'H2-Ab1', 'Cd79b', 'H2-Eb1', 'Ccna2', 'Top2a', 'Nkg7', 'Cd1d1', 'Chchd10', 'Cnp', 'Rag1',  'Rag2', 'Dntt', 'Mki67', 'Stmn1', 'Gimap8', 'Jhdm1d', 'Kif11', 'Spag5', 'Xist', 'Tsix', 'Notch2', 'Slc39a1', 'Socs3', 'Vldlr', 'Mcm4', 'Ung', 'Fam102a', 'Ccl4', 'Pyy', 'Ak4',  'Prc1', 'Cenpf', 'Prr11', 'Lhcgr', 'Tgm2', 'Abcg3', 'Ccne2', 'Pask', 'Nsg1', 'Bok', 'Klf1', 'Treml1')

FeaturePlot(subsubtiss1, genes_to_check, pt.size = 1, nCol = 3)
```


```{r, echo=FALSE, fig.height=8, fig.width=10}
#DotPlot(subsubtiss, genes_to_check, col.max = 2.5, plot.legend = T, do.return = T) + coord_flip()
```

From these genes, it appears that the clusters represent:

0: alpha
1: alpha
2: alpha
3: delta
4: alpha
5: delta
6: gamma
7: gamma

The multitude of clusters of each type correspond mostly to individual animals/sexes.

```{r}
table(FetchData(subsubtiss1, c('mouse.id','ident')) %>% droplevels())
```

```{r}
subsub.cluster.ids <- c(0, 1, 2)
subsub.free_annotation <- c(
  "HSC enriched fraction", 
  "Slamf1-negative multipotent progenitor cell", 
  "Slamf1-negative multipotent progenitor cell")
#subsub.free_annotation <- c()
subsub.cell_ontology_id = c(
  NA,
  "CL:0002035",
  "CL:0002035")
subsub.cell_ontology_class <-c(NA, "Slamf1-negative multipotent progenitor cell", "Slamf1-negative multipotent progenitor cell")

validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubtiss1@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubtiss1@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
validate_cell_ontology(subsub.cell_ontology_class)
subsub.cell_ontology_id = convert_to_cell_ontology_id(subsub.cell_ontology_class)

subsubtiss1@meta.data['free_annotation'] <- as.character(plyr::mapvalues(x = subsubtiss1@ident, from = subsub.cluster.ids, to = subsub.free_annotation))
subsubtiss1@meta.data['cell_ontology_class'] <- as.character(plyr::mapvalues(x = subsubtiss1@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_class))
subsubtiss1@meta.data['cell_ontology_id'] <- as.character(plyr::mapvalues(x = subsubtiss1@ident, from = subsub.cluster.ids, to = subsub.cell_ontology_id))
```

## Add subsubcluster cell_ontology_classs to main cell_ontology_class

```{r}
subsub.cells = rownames(subsubtiss@meta.data)

tiss@meta.data[subsub.cells, 'free_annotation'] = subsubtiss1@meta.data[,'free_annotation']
tiss@meta.data[subsub.cells, 'cell_ontology_class'] = subsubtiss1@meta.data[,'cell_ontology_class']
tiss@meta.data[subsub.cells, 'cell_ontology_id'] = subsubtiss1@meta.data[,'cell_ontology_id']
```


## Checking for batch effects
plate.barcode,mouse.id,tissue,subtissue,FACS.selection,mouse.sex
D042044,3_9_M,Marrow,B-cells,Multiple,M
D042467,3_9_M,Marrow,T-cells,Multiple,M
D041912,3_8_M,Marrow,Granulocytes,Multiple,M
D042193,3_8_M,Marrow,T-cells,Multiple,M
D042479,3_8_M,Marrow,B-cells,Multiple,M
MAA000409,3_10_M,Marrow,Granulocytes,Multiple,M
MAA000596,3_9_M,Marrow,KLS,Multiple,M
MAA000600,3_8_M,Marrow,KLS,Multiple,M
MAA000639,3_9_M,Marrow,Granulocytes,Multiple,M
MAA000652,3_10_M,Marrow,B-cells,Multiple,M
MAA000844,3_10_M,Marrow,T-cells,Multiple,M
MAA000848,3_10_M,Marrow,KLS,Multiple,M
D045139,3_39_F,Marrow,Granulocytes,Multiple,F
D045140,3_38_F,Marrow,Granulocytes,Multiple,F
MAA001842,3_38_F,Marrow,T-cells,Multiple,F
MAA001844,3_38_F,Marrow,KLS,Multiple,F
MAA001883,3_39_F,Marrow,B-cells,Multiple,F
MAA001884,3_38_F,Marrow,B-cells,Multiple,F
MAA001887,3_39_F,Marrow,KLS,Multiple,F
MAA001888,3_39_F,Marrow,T-cells,Multiple,F

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = subsubtiss, do.return = TRUE, group.by = "plate.barcode")
```

```{r}
TSNEPlot(object = subsubtiss1, do.return = TRUE, group.by = "plate.barcode")
```


# Final coloring

Color by cell ontology class on the original TSNE.

```{r}
TSNEPlot(object = subsubtiss1, do.label = TRUE, do.return = TRUE, group.by = "free_annotation")
```

# Save the Robject for later

```{r}
filename = here('00_data_ingest', '04_tissue_robj_generated', 
                     paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
print(filename)
save(tiss, file=filename)
```

```{r}
# To reload a saved object
# filename = here('00_data_ingest', '04_tissue_robj_generated', 
#                      paste0("facs", tissue_of_interest, "_seurat_tiss.Robj"))
# load(file=filename)
```


# Export the final metadata

So that Biohub can easily combine all your cell_ontology_classs, please export them as a simple csv.

```{r}
head(tiss@meta.data)
```

```{r}
filename = here('00_data_ingest', '03_tissue_cell_ontology_class_csv', 
                     paste0(tissue_of_interest, "_cell_ontology_class.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```

```{r}
filename = here('00_data_ingest', '03_tissue_annotation_csv', 
                     paste0(tissue_of_interest, "_facs_annotation.csv"))
write.csv(FetchData(tiss, c('plate.barcode','cell_ontology_class','cell_ontology_id', 'free_annotation', 'tSNE_1', 'tSNE_2')), file=filename)
```



